File: apps/desktop/src/features/wholeBook/contracts/keys.ts
/** Canonical whole-book module keys — must match backend WholeBookModuleKey. */

export const WHOLE_BOOK_MODULE_KEYS = [
  "book_overview",
  "structure_stages",
  "chapter_functions",
  "storylines",
  "characters",
  "character_arcs",
  "relationships",
  "hooks_payoffs",
  "causal_chain",
  "basic_timeline",
  "diagnostics",
] as const;

export type WholeBookModuleKey = (typeof WHOLE_BOOK_MODULE_KEYS)[number];

export const WHOLE_BOOK_ANALYSIS_MODES = [
  "whole_book_native",
  "whole_book_enhanced",
] as const;

export type WholeBookAnalysisMode = (typeof WHOLE_BOOK_ANALYSIS_MODES)[number];

export const WHOLE_BOOK_RUN_VIEW_STATUSES = [
  "pending",
  "running",
  "paused",
  "interrupted",
  "completed",
  "failed",
  "cancelled",
] as const;

export type WholeBookRunViewStatus = (typeof WHOLE_BOOK_RUN_VIEW_STATUSES)[number];

export const WHOLE_BOOK_MODULE_STATUSES = [
  "not_requested",
  "pending",
  "running",
  "partial",
  "completed",
  "failed",
  "stale",
  "blocked",
] as const;

export type WholeBookModuleStatus = (typeof WHOLE_BOOK_MODULE_STATUSES)[number];

export const STAGE_STATUSES = [
  "pending",
  "running",
  "paused",
  "interrupted",
  "completed",
  "failed",
  "skipped",
  "cancelled",
] as const;

export type StageStatus = (typeof STAGE_STATUSES)[number];

export const RUN_ALLOWED_ACTIONS = [
  "pause",
  "resume",
  "retry",
  "cancel",
  "view_partial_results",
] as const;

export type RunAllowedAction = (typeof RUN_ALLOWED_ACTIONS)[number];

export const EVIDENCE_INTEGRITY_STATUSES = [
  "valid",
  "stale",
  "hash_mismatch",
  "missing",
  "inaccessible",
] as const;

export type EvidenceIntegrityStatus = (typeof EVIDENCE_INTEGRITY_STATUSES)[number];

export const NARRATIVE_REVIEW_ACTIONS = [
  "confirm",
  "correct",
  "reject",
  "lock",
  "unlock",
  "mark_stale",
  "resolve_conflict",
  "dismiss_conflict",
] as const;

export type NarrativeReviewAction = (typeof NARRATIVE_REVIEW_ACTIONS)[number];

export const STRUCTURE_MAP_VIEW_MODES = [
  "structure_stages",
  "storylines",
  "character_growth",
] as const;

export type StructureMapViewMode = (typeof STRUCTURE_MAP_VIEW_MODES)[number];

export const RESULT_NAV_SECTION_KEYS = [
  "overview",
  "structure",
  "storylines",
  "characters",
  "relationships",
  "hooks_payoffs",
  "causal_timeline",
  "diagnostics",
  "evidence_conflicts",
  "structure_map",
] as const;

export type ResultNavSectionKey = (typeof RESULT_NAV_SECTION_KEYS)[number];

/** Product-frozen module → required engine stages (not UI stage keys). */
export const MODULE_STAGE_DEPENDENCIES: Record<
  WholeBookModuleKey,
  readonly string[]
> = {
  book_overview: ["build_fulltext_index"],
  structure_stages: [
    "build_fulltext_index",
    "resolve_entities",
    "analyze_structure",
  ],
  chapter_functions: ["analyze_structure"],
  storylines: [
    "resolve_entities",
    "analyze_structure",
    "analyze_storylines",
  ],
  characters: ["resolve_entities", "analyze_characters"],
  character_arcs: [
    "resolve_entities",
    "analyze_characters",
    "analyze_storylines",
  ],
  relationships: [
    "resolve_entities",
    "analyze_characters",
    "analyze_storylines",
  ],
  hooks_payoffs: [
    "analyze_structure",
    "analyze_storylines",
    "analyze_hooks",
  ],
  causal_chain: ["analyze_storylines", "analyze_causality_timeline"],
  basic_timeline: ["resolve_entities", "analyze_causality_timeline"],
  diagnostics: [
    "analyze_structure",
    "analyze_storylines",
    "analyze_characters",
    "analyze_hooks",
    "analyze_causality_timeline",
    "verify_evidence",
  ],
};

export const STRUCTURE_MAP_DEFAULT_MAX_NODES = 100;
export const STRUCTURE_MAP_DEFAULT_MAX_EDGES = 250;
export const PATTERN_DTO_HAS_ORM_TABLE = false;

File: apps/api/app/narrative_core/product_contract/keys.py
"""Shared product keys and module→stage dependency contract (Phase 1D-P / 1D Integration).

UI module keys must never be used as Engine Stage keys.
Module ↔ Stage is many-to-many; resolution auto-closes required stages.

Two intentional mappings (do not collapse):
- ENGINE_MODULE_PLANNING_STAGES (whole_book_stage_plan) — execution plan seeds
- PRODUCT_MODULE_STAGE_DEPENDENCIES (this module) — result status / viewability
"""

from __future__ import annotations

from collections.abc import Iterable, Sequence

from app.narrative_core.enums import WholeBookModuleKey, WholeBookStageKey
from app.narrative_core.product_contract.enums import ResultNavSectionKey

WHOLE_BOOK_MODULE_KEYS: tuple[WholeBookModuleKey, ...] = tuple(WholeBookModuleKey)

# Product Result Dependencies — which stages gate module result status / Envelope.
# Must remain a subset of Engine Planning Closure (see mapping_consistency).
PRODUCT_MODULE_STAGE_DEPENDENCIES: dict[WholeBookModuleKey, tuple[WholeBookStageKey, ...]] = {
    WholeBookModuleKey.BOOK_OVERVIEW: (
        WholeBookStageKey.BUILD_FULLTEXT_INDEX,
    ),
    WholeBookModuleKey.STRUCTURE_STAGES: (
        WholeBookStageKey.BUILD_FULLTEXT_INDEX,
        WholeBookStageKey.RESOLVE_ENTITIES,
        WholeBookStageKey.ANALYZE_STRUCTURE,
    ),
    WholeBookModuleKey.CHAPTER_FUNCTIONS: (
        WholeBookStageKey.ANALYZE_STRUCTURE,
    ),
    WholeBookModuleKey.STORYLINES: (
        WholeBookStageKey.RESOLVE_ENTITIES,
        WholeBookStageKey.ANALYZE_STRUCTURE,
        WholeBookStageKey.ANALYZE_STORYLINES,
    ),
    WholeBookModuleKey.CHARACTERS: (
        WholeBookStageKey.RESOLVE_ENTITIES,
        WholeBookStageKey.ANALYZE_CHARACTERS,
    ),
    WholeBookModuleKey.CHARACTER_ARCS: (
        WholeBookStageKey.RESOLVE_ENTITIES,
        WholeBookStageKey.ANALYZE_CHARACTERS,
        WholeBookStageKey.ANALYZE_STORYLINES,
    ),
    WholeBookModuleKey.RELATIONSHIPS: (
        WholeBookStageKey.RESOLVE_ENTITIES,
        WholeBookStageKey.ANALYZE_CHARACTERS,
        WholeBookStageKey.ANALYZE_STORYLINES,
    ),
    WholeBookModuleKey.HOOKS_PAYOFFS: (
        WholeBookStageKey.ANALYZE_STRUCTURE,
        WholeBookStageKey.ANALYZE_STORYLINES,
        WholeBookStageKey.ANALYZE_HOOKS,
    ),
    WholeBookModuleKey.CAUSAL_CHAIN: (
        WholeBookStageKey.ANALYZE_STORYLINES,
        WholeBookStageKey.ANALYZE_CAUSALITY_TIMELINE,
    ),
    WholeBookModuleKey.BASIC_TIMELINE: (
        WholeBookStageKey.RESOLVE_ENTITIES,
        WholeBookStageKey.ANALYZE_CAUSALITY_TIMELINE,
    ),
    WholeBookModuleKey.DIAGNOSTICS: (
        WholeBookStageKey.ANALYZE_STRUCTURE,
        WholeBookStageKey.ANALYZE_STORYLINES,
        WholeBookStageKey.ANALYZE_CHARACTERS,
        WholeBookStageKey.ANALYZE_HOOKS,
        WholeBookStageKey.ANALYZE_CAUSALITY_TIMELINE,
        WholeBookStageKey.VERIFY_EVIDENCE,
    ),
}

# Deprecated alias — prefer PRODUCT_MODULE_STAGE_DEPENDENCIES.
MODULE_STAGE_DEPENDENCIES = PRODUCT_MODULE_STAGE_DEPENDENCIES

# Result page navigation (product IA). Module keys map into sections.
RESULT_NAV_SECTIONS: tuple[tuple[ResultNavSectionKey, tuple[WholeBookModuleKey, ...]], ...] = (
    (ResultNavSectionKey.OVERVIEW, (WholeBookModuleKey.BOOK_OVERVIEW,)),
    (
        ResultNavSectionKey.STRUCTURE,
        (WholeBookModuleKey.STRUCTURE_STAGES, WholeBookModuleKey.CHAPTER_FUNCTIONS),
    ),
    (ResultNavSectionKey.STORYLINES, (WholeBookModuleKey.STORYLINES,)),
    (
        ResultNavSectionKey.CHARACTERS,
        (WholeBookModuleKey.CHARACTERS, WholeBookModuleKey.CHARACTER_ARCS),
    ),
    (ResultNavSectionKey.RELATIONSHIPS, (WholeBookModuleKey.RELATIONSHIPS,)),
    (ResultNavSectionKey.HOOKS_PAYOFFS, (WholeBookModuleKey.HOOKS_PAYOFFS,)),
    (
        ResultNavSectionKey.CAUSAL_TIMELINE,
        (WholeBookModuleKey.CAUSAL_CHAIN, WholeBookModuleKey.BASIC_TIMELINE),
    ),
    (ResultNavSectionKey.DIAGNOSTICS, (WholeBookModuleKey.DIAGNOSTICS,)),
    (ResultNavSectionKey.EVIDENCE_CONFLICTS, ()),
    (ResultNavSectionKey.STRUCTURE_MAP, ()),
)

FUTURE_API_ROUTES: tuple[str, ...] = (
    "POST /api/v1/books/{book_id}/whole-book-runs",
    "GET /api/v1/whole-book-runs/{run_id}",
    "GET /api/v1/whole-book-runs/{run_id}/stages",
    "POST /api/v1/whole-book-runs/{run_id}/pause",
    "POST /api/v1/whole-book-runs/{run_id}/resume",
    "POST /api/v1/whole-book-runs/{run_id}/cancel",
    "POST /api/v1/whole-book-runs/{run_id}/stages/{stage_key}/retry",
    "GET /api/v1/narrative-assets/{asset_id}/evidence",
    "POST /api/v1/narrative-review-actions",
    "GET /api/v1/books/{book_id}/analysis-conflicts",
)

EXISTING_API_ROUTES: tuple[str, ...] = (
    "GET /api/v1/capabilities",
    "GET /api/v1/capabilities/{key}",
    "POST /api/v1/books/{book_id}/whole-book-runs/preflight",
    "GET /api/v1/whole-book-runs/{run_id}/results",
    "GET /api/v1/whole-book-runs/{run_id}/results/{module_key}",
)


def _normalize_module(raw: WholeBookModuleKey | str) -> WholeBookModuleKey:
    if isinstance(raw, WholeBookModuleKey):
        return raw
    return WholeBookModuleKey(str(raw))


def resolve_modules_with_dependencies(
    requested: Sequence[WholeBookModuleKey | str],
) -> tuple[tuple[WholeBookModuleKey, ...], tuple[WholeBookStageKey, ...], tuple[str, ...]]:
    """Resolve requested modules → unique modules + required stages + user-facing notes.

    Users cannot cancel required dependency stages. Auto-filled stages are explained
    via ``auto_fill_notes``.
    """

    if not requested:
        modules = tuple(WHOLE_BOOK_MODULE_KEYS)
    else:
        seen: list[WholeBookModuleKey] = []
        for raw in requested:
            module = _normalize_module(raw)
            if module not in seen:
                seen.append(module)
        modules = tuple(seen)

    stages: list[WholeBookStageKey] = []
    for module in modules:
        for stage in PRODUCT_MODULE_STAGE_DEPENDENCIES[module]:
            if stage not in stages:
                stages.append(stage)

    notes: list[str] = []
    for module in modules:
        deps = PRODUCT_MODULE_STAGE_DEPENDENCIES[module]
        notes.append(
            f"module {module.value} requires stages: "
            + ", ".join(s.value for s in deps)
        )
    return modules, tuple(stages), tuple(notes)


def stages_for_modules(modules: Iterable[WholeBookModuleKey | str]) -> tuple[WholeBookStageKey, ...]:
    _, stages, _ = resolve_modules_with_dependencies(tuple(modules))
    return stages
